### PR TITLE
HTTP ingress rule value is Mandatory

### PR DESCRIPTION
**Pull request by kuritka**
related to #401 

- the custom structures with mandatory field overrides `IngressSpec` and
and implements necessary routines like `DeepCopyInto` and transformations between
`k8gbv1beta1` and `v1beta1`

- controller-gen created `DeepCopy()` functions in `zz_generated.deepcopy.go`

### CODE REVIEW

File: terratest/test/k8gb_ingress_mandatory_http_test.go [hunk 1]
```diff
@@ -0,0 +1,65 @@
+/*
```
Comment (k0da) [inline terratest/test/k8gb_ingress_mandatory_http_test.go:1]: We already have it covered in terratest/test/k8gb_basic_app_test.go

Comment (kuritka) [reply terratest/test/k8gb_ingress_mandatory_http_test.go:1]: @k0da , I see, amending....

**Comment by kuritka** [comment]
@k0da amended

**Review by ytsarev — CHANGES_REQUESTED** [review]
From https://kubernetes.io/blog/2019/07/18/api-deprecations-in-1-16/

```
The v1.22 release will stop serving the following deprecated API versions in favor of newer and more stable API versions:

Ingress in the extensions/v1beta1 API version will no longer be served
Migrate to use the networking.k8s.io/v1beta1 API version, available since v1.14. Existing persisted data can be retrieved/updated via the new version.
```

So we should stick with `networking.k8s.io/v1beta1 `

**Comment by kuritka** [comment]
@ytsarev, @k0da  - so should we upgrade to `networking.k8s.io/v1beta1 ` first and then migrate this PR to `networking.k8s.io/v1beta1 `??

**Comment by k0da** [comment]
@kuritka IMO we can do both

**Comment by kuritka** [comment]
@ytsarev @k0da @somaritane , I succesfully migrated to `networking.k8s.io/v1beta1` (see: [this](https://github.com/AbsaOSS/k8gb/compare/ingress-mandatory-http-networking-v1beta1)), but to make PR slim I rather split large one into into two smaller PR's. 

This one is making `HTTPIngressRuleValue` mandatory. The migration would go to the second PR.

just in case: 
To migrate into `networking.k8s.io/v1beta1`, I had to [change CRD](https://github.com/AbsaOSS/k8gb/compare/ingress-mandatory-http-networking-v1beta1#diff-b89a907c86922705f2e446b83056186485868f3152f770c9b36e5ed675fecd83R221) and `roles.yaml` (would be second PR).

**Review by k0da — DISMISSED** [review]
👍

File: api/v1beta1/types.go [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+	if in.Rules != nil {
+		in, out := &in.Rules, &out.Rules
+		*out = make([]IngressRule, len(*in))
+		for i := range *in {
+			(*in)[i].DeepCopyInto(&(*out)[i])
+		}
+	}
+}
+
+// DeepCopyInto is an autogenerated deepcopy function, copying the receiver, writing into out. in must be non-nil.
```
Comment (somaritane) [inline api/v1beta1/types.go:146]: do we autogenerate this function here too, like in zz_generated.deepcopy.go?

Comment (kuritka) [reply api/v1beta1/types.go:146]: These are not auto-generated, amending comments.

Comment (kuritka) [reply api/v1beta1/types.go]: @somaritane comments amended

File: api/v1beta1/types.go [hunk 3]
```diff
@@ -0,0 +1,196 @@
+/*
```
Comment (ytsarev) [inline api/v1beta1/types.go:1]: What is the filename semantics? `types` vs `gslb_types`?

File: api/v1beta1/types.go [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+
+Generated by GoLic, for more details see: https://github.com/AbsaOSS/golic
+*/
+package v1beta1
+
+import (
+	"k8s.io/api/extensions/v1beta1"
+)
+
+type IngressSpec struct {
```
Comment (ytsarev) [inline api/v1beta1/types.go:24]: Can we reference the upstream source of `IngressSpec` for future?

**Review by ytsarev — CHANGES_REQUESTED** [review]
some cosmetic comments with the focus on future maintenance

File: api/v1beta1/types.go (hunk 3 above)
Comment (kuritka) [reply api/v1beta1/types.go:1]: @ytsarev  good question, 
It lives originally in `k8s.io/api/extensions/v1beta1/types.go`.

File: api/v1beta1/types.go (hunk 4 above)
Comment (kuritka) [reply api/v1beta1/types.go:24]: @ytsarev , good idea, amended

File: api/v1beta1/types.go (hunk 3 above)
Comment (ytsarev) [reply api/v1beta1/types.go:1]: `upstream_types.go` maybe?

File: api/v1beta1/types.go [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+Generated by GoLic, for more details see: https://github.com/AbsaOSS/golic
+*/
+package v1beta1
+
+import (
+	"k8s.io/api/extensions/v1beta1"
+)
+
+// IngressSpec overrides https://github.com/kubernetes/api/blob/master/extensions/v1beta1/types.go
+// IngressSpec describes the Ingress the user wishes to exist.
```
Comment (somaritane) [inline api/v1beta1/types.go:25]: This comment probably needs rephrasing.

Comment (ytsarev) [inline api/v1beta1/types.go:25]:
```suggestion
// IngressSpec is upstream Ingress to be included in Gslb specification and extended with Gslb specific requirements.
```

**Review by ytsarev — CHANGES_REQUESTED** [review]
IngressSpec Wording suggestion.

File: api/v1beta1/types.go (hunk 3 above)
Comment (kuritka) [reply api/v1beta1/types.go]: Renamed, was just following original convention `upstream_types.go` is well descriptive.

**Review by ytsarev — APPROVED** [review]

**Review by somaritane — APPROVED** [review]
lgtm

**Pull request merged**